src/tests/suite/daos_dist_tx.c

```diff
@@ -0,0 +1,986 @@
+
+	req.arg->expect_result = 0;
+	punch_akey_with_flags(dkey, akey, th, &req, DAOS_COND_PUNCH);
+
+	req.arg->expect_result = -DER_NONEXIST;
+	punch_dkey_with_flags(dkey2, th, &req, DAOS_COND_PUNCH);
+
+	req.arg->expect_result = 0;
+	punch_dkey_with_flags(dkey, th, &req, DAOS_COND_PUNCH);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	assert_int_equal(req.iod[0].iod_size, 0);
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_11(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*dkey2 = "tmp_dkey";
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+
+	print_message("Read only transaction\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_SINGLE, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	insert_single(dkey2, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	MUST(daos_tx_open(arg->coh, &th, DAOS_TF_RDONLY, NULL));
+
+	req.arg->expect_result = -DER_NO_PERM;
+
+	insert_single(dkey, akey2, 0, write_buf, DTX_IO_SMALL, th, &req);
+	punch_akey(dkey, akey, th, &req);
+
+	req.arg->expect_result = 0;
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, DTX_IO_SMALL);
+
+	lookup_single(dkey, akey2, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, 0);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_12(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 i;
+
+	print_message("DTX with zero copy flag\n");
+	MUST(daos_tx_open(arg->coh, &th, DAOS_TF_ZERO_COPY, NULL));
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	/* Reuse 'write_buf' */
+	for (i = 0; i < DTX_IO_SMALL; i++)
+		write_buf[i] += 1;
+
+	insert_single(dkey, akey2, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	lookup_single(dkey, akey2, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	assert_memory_equal(write_buf, fetch_buf, DTX_IO_SMALL);
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	/* 'write_buf' has been overwritten, so it is the same as 2nd update. */
```

Comment:
Hm... that is only used for verify whether our ZERO_COPY flags works or not. If works, the old buffer will be overwritten; otherwise, if not overwritten, then means we copied the data unexpectedly.

From the API perspective, I agree with you that we should not promise to the caller for the heavier of reusing buffer with ZERO_COPY flag.